infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/pvc/UniqueWorkspacePVCStrategy.java

```diff
@@ -59,41 +61,50 @@ public UniqueWorkspacePVCStrategy(
       @Named("che.infra.openshift.pvc.name") String pvcName,
       @Named("che.infra.openshift.pvc.quantity") String pvcQuantity,
       @Named("che.infra.openshift.pvc.access_mode") String pvcAccessMode,
-      @Named("che.workspace.projects.storage") String projectFolderPath,
       OpenShiftClientFactory clientFactory,
       OpenShiftProjectFactory factory) {
     this.pvcName = pvcName;
     this.pvcQuantity = pvcQuantity;
     this.projectName = projectName;
     this.pvcAccessMode = pvcAccessMode;
-    this.projectsPath = projectFolderPath;
     this.clientFactory = clientFactory;
     this.factory = factory;
   }
 
   @Override
   public void prepare(OpenShiftEnvironment osEnv, String workspaceId)
       throws InfrastructureException {
-    final String machineWithSources =
-        getWsAgentServerMachine(osEnv)
-            .orElseThrow(() -> new InfrastructureException("Machine with ws-agent not found"));
-    final String pvcUniqueName = pvcName + '-' + workspaceId;
-    final Map<String, PersistentVolumeClaim> claims = osEnv.getPersistentVolumeClaims();
-    claims.put(pvcUniqueName, newPVC(pvcUniqueName, pvcAccessMode, pvcQuantity));
-    factory.create(workspaceId).persistentVolumeClaims().createIfNotExist(claims.values());
+    Map<String, PersistentVolumeClaim> claims = osEnv.getPersistentVolumeClaims();
     for (Pod pod : osEnv.getPods().values()) {
       final PodSpec podSpec = pod.getSpec();
       for (Container container : podSpec.getContainers()) {
         final String machineName = Names.machineName(pod, container);
-        if (machineName.equals(machineWithSources)) {
-          final String subPath =
-              projectsPath.startsWith("/") ? projectsPath.substring(1) : projectsPath;
-          container.getVolumeMounts().add(newVolumeMount(pvcUniqueName, projectsPath, subPath));
-          podSpec.getVolumes().add(newVolume(pvcUniqueName, pvcUniqueName));
-          return;
-        }
+        InternalMachineConfig machineConfig = osEnv.getMachines().get(machineName);
+        addMachineVolumes(workspaceId, claims, podSpec, container, machineConfig);
       }
     }
+    factory.create(workspaceId).persistentVolumeClaims().createIfNotExist(claims.values());
+  }
+
+  private void addMachineVolumes(
+      String workspaceId,
+      Map<String, PersistentVolumeClaim> claims,
+      PodSpec podSpec,
+      Container container,
+      InternalMachineConfig machineConfig) {
+    if (machineConfig == null || machineConfig.getVolumes().isEmpty()) {
+      return;
+    }
+    for (Entry<String, Volume> volumeEntry : machineConfig.getVolumes().entrySet()) {
+      String volumeName = volumeEntry.getKey();
+      String volumePath = volumeEntry.getValue().getPath();
+      String subPath = workspaceId + '/' + volumeName;
+      String pvcUniqueName = pvcName + '-' + workspaceId + '-' + volumeName;
+
+      container.getVolumeMounts().add(newVolumeMount(pvcUniqueName, volumePath, subPath));
```

Comment:
Should you add volume mount only if it doesn't exist? Or mb rewrite existing one